### PR TITLE
Implement Test and Validation Loss for Flux Finetuning and LoRA Training

### PR DESCRIPTION
**Pull request by stepfunction83**
As discussed in the issue I created here: https://github.com/kohya-ss/sd-scripts/issues/1897

This is based off of the work done by spacepxl here: https://github.com/spacepxl/demystifying-sd-finetuning

The general point of it is that traditionally, the loss that is calculated at each step is more or less fully random and conveys very little information about the actual progress of training and very little about the performance. 

Instead, we can calculate test and validation loss values as one would when doing traditional machine learning. By selecting a specific set of samples and running them with identical noise and timesteps over time, we can evaluate a test loss for the model. Further, by selecting a portion of the training set to hold out, we can calculate a validation error as well as identify when the model begins overtraining.

This means that until this capability becomes available, we are effectively just guessing at the progress of our training and have no real quantitative measurement of progress or of overtraining.

In this PR, I implement this capability for Flex Finetuning and LoRA training which allows for these two sets to be automatically identified and for the two loss values to be captured at regular intervals.

So far, I've done several runs to verify that it is working properly, and you can see a brief snippet of this here:
image
(Pictured: Test loss relative to initial value, Validation loss relative to initial value)

The test error relative to the initial test error is on the left, while the respective validation results are on the right. As opposed to the raw loss values (below), you can see they are much smoother and more directionally indicative of training progress.
image
(Pictured: Raw loss values, i.e. garbage)

To use the functionality, I have added 5 new arguments:
--test_step_freq 25 (The frequency at which test loss is calculated; 0 turns off feature)
--val_step_freq 25 (The frequency at which validation loss is calculated; 0 turns off feature)
--test_set_count 10 (The number of training samples used for the test set)
--val_set_count 10 (The number of training samples used for the validation set; These are then skipped in training since they are held out)
--test_val_repeat_count 6 (The number of noise/timestep iterations to use for each test and validation sample)

I will say that this required some fairly sizable changes to the training loops to accommodate the additional loss calculations seamlessly, so I would definitely like feedback regarding the approach. I'm also not very familiar with accelerate and pytorch (and am a middling programmer to boot), so I may be doing something wrong along the way and almost certainly could benefit from some best practices. From my initial tests however, things at least seem to be working appropriately and generating the desired result.

This is also limited to just Flux for now, but the general logic that I used could be applied more broadly to other training types.

### CODE REVIEW

**Comment by stepfunction83** [comment]
> Thank you for this! Unfortunately, it seems to be a duplicate of #1864. However, your approach for `calculate_loss` is very simple and great. I think we can simplify the code this way in the future.
> 
> However, I think there is one drawback to using the top n items of the dataset. That is, the data returned by the dataloader is shuffled every time. Therefore, if I understand this PR correctly, the current code will be testing/validating the same data as the training data, and the correct loss will not be obtained.
> 
> To address this issue, we need to have an independent validation dataset, which is what we do in #1864.

So in this, I am calculating loss for both a test and a validation set. Both are selected from the train_dataloader, and both are tested using a consistent set of timesteps and noise over time to ensure consistent loss calculation. The primary difference is that for the validation set, the train loop automatically skips over those samples when it encounters them in training, so they do not affect the final model. You can see that here at the beginning of the train loop:

```
if args.val_step_freq > 0:
                if step in val_set['steps']: # Skip validation steps, don't increment global step
                    logger.warning('SKIPPING BATCH IN VALIDATION SET')
                    continue
```

**Comment by stepfunction83** [comment]
I'll close this for now and begin migrating some of the ideas into another PR using the new framework.
